sfge/src/main/java/com/salesforce/telemetry/TelemetryUtil.java

```diff
@@ -0,0 +1,161 @@
+package com.salesforce.telemetry;
+
+import com.google.gson.Gson;
+import com.salesforce.exception.SfgeRuntimeException;
+import com.salesforce.graph.symbols.apex.ApexValue;
+import com.salesforce.graph.vertex.InvocableVertex;
+import com.salesforce.graph.vertex.SFVertex;
+import com.salesforce.messaging.CliMessager;
+import com.salesforce.messaging.EventKey;
+import java.util.Arrays;
+import java.util.stream.Collectors;
+import org.apache.logging.log4j.Logger;
+
+public final class TelemetryUtil {
+    private static final String MAIN_THREAD_NAME = "main";
+
+    public static void postTelemetryAndLogWarning(Logger logger, String msg) {
+        TelemetryData telemetryData = new TelemetryData(msg, EventType.LOG_WARNING);
+        CliMessager.postMessage(
+                "TelemetryData", EventKey.INFO_TELEMETRY, new Gson().toJson(telemetryData));
+        if (logger.isWarnEnabled()) {
+            logger.warn(msg);
+        }
+    }
+
+    public static void postTelemetryAndLogWarning(Logger logger, String msg, SFVertex vertex) {
+        TelemetryData telemetryData = new TelemetryData(msg, EventType.LOG_WARNING, vertex);
+        CliMessager.postMessage(
+                "TelemetryData", EventKey.INFO_TELEMETRY, new Gson().toJson(telemetryData));
+        if (logger.isWarnEnabled()) {
+            logger.warn(msg);
+        }
+    }
+
+    public static void postTelemetryAndLogWarning(
+            Logger logger, String msg, InvocableVertex vertex) {
+        TelemetryData telemetryData = new TelemetryData(msg, EventType.LOG_WARNING, vertex);
+        CliMessager.postMessage(
+                "TelemetryData", EventKey.INFO_TELEMETRY, new Gson().toJson(telemetryData));
+        if (logger.isWarnEnabled()) {
+            logger.warn(msg);
+        }
+    }
+
+    public static void postTelemetryAndThrow(SfgeRuntimeException runtimeException) {
+        // This is predicated on the assumption that the main thread's name will never be changed
```

Comment:
The alternative would be to immediately cache the main thread's ID at the start of `main` so we can compare it here. I think that this implementation is acceptable for now.